sycl/include/sycl/ext/intel/experimental/esimd/common.hpp

```diff
@@ -176,11 +176,279 @@ enum class atomic_op : uint8_t {
+                Op == atomic_op::fsub || Op == atomic_op::fmin ||
+                Op == atomic_op::fmax || Op == atomic_op::bit_and ||
+                Op == atomic_op::bit_or || Op == atomic_op::bit_xor) {
+    static_assert(NumSrc == 1, "One source operand is expected");
+  }
+  if constexpr (Op == atomic_op::cmpxchg || Op == atomic_op::fcmpwr) {
+    static_assert(NumSrc == 2, "Two source operands are expected");
+  }
+}
+
+template <atomic_op Op> constexpr lsc_atomic_op to_lsc_atomic_op() {
+  check_lsc_atomic_op<Op>();
+  switch (Op) {
+  case atomic_op::add:
+    return lsc_atomic_op::iadd;
+  case atomic_op::sub:
+    return lsc_atomic_op::isub;
+  case atomic_op::inc:
+    return lsc_atomic_op::iinc;
+  case atomic_op::dec:
+    return lsc_atomic_op::idec;
+  case atomic_op::min:
+    return lsc_atomic_op::umin;
+  case atomic_op::max:
+    return lsc_atomic_op::umax;
+  case atomic_op::cmpxchg:
+    return lsc_atomic_op::icas;
+  case atomic_op::bit_and:
+    return lsc_atomic_op::bit_and;
+  case atomic_op::bit_or:
+    return lsc_atomic_op::bit_or;
+  case atomic_op::bit_xor:
+    return lsc_atomic_op::bit_xor;
+  case atomic_op::minsint:
+    return lsc_atomic_op::smin;
+  case atomic_op::maxsint:
+    return lsc_atomic_op::smax;
+  case atomic_op::fmax:
+    return lsc_atomic_op::fmax;
+  case atomic_op::fmin:
+    return lsc_atomic_op::fmin;
+  case atomic_op::fcmpwr:
+    return lsc_atomic_op::fcas;
+  case atomic_op::fadd:
+    return lsc_atomic_op::fadd;
+  case atomic_op::fsub:
+    return lsc_atomic_op::fsub;
+  case atomic_op::load:
+    return lsc_atomic_op::load;
+  case atomic_op::store:
+    return lsc_atomic_op::store;
+  default:
+    return lsc_atomic_op::iinc;
+  }
+}
+
+template <lsc_vector_size VS> constexpr uint8_t to_int() {
+  check_lsc_vector_size<VS>();
+  switch (VS) {
+  case lsc_vector_size::n1:
+    return 1;
+  case lsc_vector_size::n2:
+    return 2;
+  case lsc_vector_size::n3:
+    return 3;
+  case lsc_vector_size::n4:
+    return 4;
+  case lsc_vector_size::n8:
+    return 8;
+  case lsc_vector_size::n16:
+    return 16;
+  case lsc_vector_size::n32:
+    return 32;
+  case lsc_vector_size::n64:
+    return 64;
+  default:
+    return 1;
+  }
+}
+
+template <uint8_t VS> constexpr lsc_vector_size to_lsc_vector_size() {
+  check_lsc_vector_size<VS>();
+  switch (VS) {
+  case 1:
+    return lsc_vector_size::n1;
+  case 2:
+    return lsc_vector_size::n2;
+  case 3:
+    return lsc_vector_size::n3;
+  case 4:
+    return lsc_vector_size::n4;
+  case 8:
+    return lsc_vector_size::n8;
+  case 16:
+    return lsc_vector_size::n16;
+  case 32:
+    return lsc_vector_size::n32;
+  case 64:
+    return lsc_vector_size::n64;
+  default:
+    return lsc_vector_size::n1;
+  }
+}
+
+template <typename T, lsc_data_size DS>
+constexpr lsc_data_size finalize_data_size() {
+  check_lsc_data_size<T, DS>();
+  if (DS != lsc_data_size::default_size)
+    return DS;
+  else if (sizeof(T) == 1)
```

Comment:
Actually U8 and U16 data size values are only supported by 2d block messages. D32U8 and D32U16 should be used instead

Reply:
I think this has been fixed in https://github.com/intel/llvm/pull/5512/commits/f922b01588bc7888c5635d0c4195daeeeeeac872

Reply:
Probably the fixed code will not work. When `sizeof(T)` is 1 or 2, values should be also bit-casted to `uint32_t`.

Reply:
Please review the updated patch. I think this problem has been fixed.